File: my_py_actions/my_py_actions/count_until_client.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle, GoalStatus

from my_robot_interfaces.action import CountUntil


class CountUntilClientNode(Node):
    def __init__(self):
        super().__init__("count_until_client")

        self.count_until_client_ = ActionClient(
            node=self, action_type=CountUntil, action_name="count_until"
        )

        self.get_logger().info("Count Until Client started.")

    def send_goal(self, target, period):
        self.count_until_client_.wait_for_server()

        goal = CountUntil.Goal()
        goal.target = target
        goal.period = period

        self.get_logger().info("Sending goal.")
        self.count_until_client_.send_goal_async(
            goal, feedback_callback=self.goal_feedback_callback
        ).add_done_callback(self.goal_response_callback)

        # self.timer_ = self.create_timer(2.0, self.cancel_goal)

    # def cancel_goal(self):
    #     self.get_logger().info("Sending cancel request.")
    #     self.goal_handle_.cancel_goal_async()
    #     self.timer_.cancel()

    def goal_response_callback(self, future):
        self.goal_handle_: ClientGoalHandle = future.result()
        if self.goal_handle_.accepted:
            self.get_logger().info("Goal accepted.")
            self.goal_handle_.get_result_async().add_done_callback(
                self.goal_result_callback
            )
        else:
            self.get_logger().warn("Goal rejected.")

    def goal_result_callback(self, future):
        status = future.result().status

        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info("Goal succeeded.")
        elif status == GoalStatus.STATUS_ABORTED:
            self.get_logger().error("Goal aborted.")
        elif status == GoalStatus.STATUS_CANCELED:
            self.get_logger().warn("Goal canceled.")

        result: CountUntil.Result = future.result().result
        self.get_logger().info(f"Count result is {result.count}.")

    def goal_feedback_callback(self, feedback_msg):
        feedback: CountUntil.Feedback = feedback_msg.feedback
        self.get_logger().info(f"Got feedback of count to {feedback.count}.")


def main(args=None):
    rclpy.init(args=args)

    node = CountUntilClientNode()
    node.send_goal(5, 1.0)
    rclpy.spin(node)

    rclpy.shutdown()


if __name__ == "main":
    main()

File: my_cpp_actions/src/count_until_client.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp_action/rclcpp_action.hpp>

#include <my_robot_interfaces/action/count_until.hpp>

using namespace std::placeholders;

class CountUntilClientNode : public rclcpp::Node
{
public:
  CountUntilClientNode() : Node("count_until_client")
  {
    pActionClient_ = rclcpp_action::create_client<my_robot_interfaces::action::CountUntil>(this, "count_until");

    RCLCPP_INFO(this->get_logger(), "Count Until Client started.");
  }

  void sendGoal(const int64_t target, const double period)
  {
    pActionClient_->wait_for_action_server();

    auto goal = my_robot_interfaces::action::CountUntil::Goal();
    goal.target = target;
    goal.period = period;

    auto options = rclcpp_action::Client<my_robot_interfaces::action::CountUntil>::SendGoalOptions();
    options.goal_response_callback = std::bind(&CountUntilClientNode::cbGoalResponse, this, _1);
    options.feedback_callback = std::bind(&CountUntilClientNode::cbGoalFeedback, this, _1, _2);
    options.result_callback = std::bind(&CountUntilClientNode::cbGoalResult, this, _1);

    RCLCPP_INFO(this->get_logger(), "Sending goal.");
    pActionClient_->async_send_goal(goal, options);

    // pTimer_ = this->create_wall_timer(std::chrono::seconds(2), std::bind(&CountUntilClientNode::cbCancelGoal, this));
  }

private:
  // void cbCancelGoal()
  // {
  //   RCLCPP_INFO(this->get_logger(), "Sending cancel request.");
  //   assert(pGoalHandle_);
  //
  //   pActionClient_->async_cancel_goal(pGoalHandle_);
  //
  //   pTimer_->cancel();
  // }

  void cbGoalResponse(const rclcpp_action::ClientGoalHandle<my_robot_interfaces::action::CountUntil>::SharedPtr &pGoalHandle)
  {
    if (!pGoalHandle)
    {
      RCLCPP_INFO(this->get_logger(), "Goal rejected.");
    }
    else
    {
      pGoalHandle_ = pGoalHandle;
      RCLCPP_INFO(this->get_logger(), "Goal accepted.");
    }
  }

  void cbGoalFeedback(const rclcpp_action::ClientGoalHandle<my_robot_interfaces::action::CountUntil>::SharedPtr & /*pGoalHandle*/,
                      const my_robot_interfaces::action::CountUntil::Feedback::ConstSharedPtr pFeedback)
  {
    RCLCPP_INFO(this->get_logger(), "Got feedback of count to %ld.", pFeedback->count);
  }

  void cbGoalResult(const rclcpp_action::ClientGoalHandle<my_robot_interfaces::action::CountUntil>::WrappedResult &result)
  {
    const auto status = result.code;
    if (status == rclcpp_action::ResultCode::SUCCEEDED)
    {
      RCLCPP_INFO(this->get_logger(), "Goal succeeded.");
    }
    else if (status == rclcpp_action::ResultCode::ABORTED)
    {
      RCLCPP_ERROR(this->get_logger(), "Goal aborted.");
    }
    if (status == rclcpp_action::ResultCode::CANCELED)
    {
      RCLCPP_WARN(this->get_logger(), "Goal canceled.");
    }

    const auto count = result.result->count;
    RCLCPP_INFO(this->get_logger(), "Count result is %ld.", count);
  }

  rclcpp_action::Client<my_robot_interfaces::action::CountUntil>::SharedPtr pActionClient_;
  rclcpp_action::ClientGoalHandle<my_robot_interfaces::action::CountUntil>::SharedPtr pGoalHandle_;

  // rclcpp::TimerBase::SharedPtr pTimer_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);

  auto pNode = std::make_shared<CountUntilClientNode>();
  pNode->sendGoal(8, 1);
  rclcpp::spin(pNode);

  rclcpp::shutdown();
  return 0;
}